Add Gen4 support of command 'temp'

kelvin-cao opened the pull request:
The API switchtec_die_temp() now supports both Gen3 and Gen4 devices.

Note: The subcommand will be changed in Gen4 F/W, need update later.

kelvin-cao committed:
Introduce get_device_id op instance for all interfaces

kelvin-cao committed:
Add support of device_id/gen/variant

A new device table was introduced to map device_id and gen/variant.
Currently only Gen3 PSX/PFX is supported, will remove the condition
once Gen4 or other variants are functionally supported.

kelvin-cao committed:
Introduce CLI command 'info'

To display basic info of device node.

wesleywesley committed:
Microsemi/switchtec-user is now equipped with travis-ci,
contributor can create pull request against this repo directly.
Update the "pull request" section accordingly.

Signed-off-by: Wesley Sheng <wesley.sheng@microchip.com>

wesleywesley committed:
Display the physical port when list all port informaion

When port state in unbound, bind fail, bind in progress, the physical
port information displayed when specify for a single port, but doesn't
appear when list all port information.

Fix this bug by display physical port in all the port states.

Signed-off-by: Wesley Sheng <wesley.sheng@microchip.com>

wesleywesley committed:
Merge pull request #91 from Microsemi/port_bind_info_bug

Display the physical port when list all port informaion

wesleywesley committed:
Merge pull request #86 from Microsemi/contributing_md

Update CONTRIBUTING.md file

Commit:
Introduce variable pax in struct switchtec_dev

Add variable pax for PAX id, which is only valid for the Switchtec
PAX variants. The PAX id is encoded in the command option field when
issue command.

With valid PAX id, the firmware either execute the command on local
switch if the PAX id is for local, or forward the command to remote
PAX switch. In the remote case, local firmware acts as a command agent.
The actual execution of the command is on the remote PAX.

The PAX ID can be set by the user using switchtec_set_pax_id().

Commit:
Add global pax id handling functions for CLI

The PAX ID may be set on the command line before or after the device. So
we ensure that it is set only after device has been processed. The PAX ID
is set to SWITCHTEC_PAX_ID_LOCAL, if the pax option is not specified.

kelvin-cao committed:
Add Gen4 support of command 'temp'

The API switchtec_die_temp() now supports both Gen3 and Gen4 devices.

Note: The subcommand will be changed in Gen4 F/W, need update later.

lsgunth reviewed (commented):
Looks fine to me. Merge once you've confirmed that ID value.

Though, as a side note, I think it's crazy the firmware team changed this interface. There doesn't seem to be any significant improvement and it just creates more work for us and extra code to maintain backwards compatibility. If there's no good justification the firmware team should not change interfaces like this. The existing interface is ten times better than any new one simply because it's compatible.

kelvin-cao commented:
I talked really a lot about these with firmware team, they used to think the compatiblity is nothing since users do need to update it and make some changes. I think things are changing and we have all agreed to keep the firmware interface compatible with a Gen, but I expect there still will be cases like this.

kelvin-cao committed:
Enhance switchtec error handling to make it MRPC aware

In the switchtec API functions, errno was used to store either MRPC
specific errors or other systme errors. To distinguish between these two
cases, we use the 31st bit (second highest bit) of errno to indicates
it's a MRPC error (bit set) or a system error (bit clear).

A new global variable mprc_error_cmd was also introduced to store
the MRPC command ID when the errno was for MRPC, as the firmware
don't ensure all the errno are unique across all MRPC commands.

wesleywesley committed:
Fix cfg_positive_handler function bug

cfg_positive_handler function was designed for check and parse
the positive argument, and warning when negative or zero argument
was input. But current implementation negative and zero argument
input was not filter out, because strtoul was used.
Fix it by change from strtoul to strtol, and doulbe check if the
output of strtol is negative or zero, warning accordingly.

Signed-off-by: Wesley Sheng <wesley.sheng@microchip.com>

wesleywesley committed:
Add argument type and handler for nonnegative argument

Add argument type and handler for nonnegative argument, when
the argument required to be a nonnegative value, and the
positive argument type and corresponding handler is not
applicable.

Signed-off-by: Wesley Sheng <wesley.sheng@microchip.com>

wesleywesley committed:
Replace several argument type from positive to nonnegative

Some of the arguments' type should be nonnegative instead of
positive, change it accordingly.

Signed-off-by: Wesley Sheng <wesley.sheng@microchip.com>

wesleywesley committed:
Merge pull request #94 from Microsemi/cfg_positive_handler_fix

Cfg positive handler fix

wesleywesley committed:
Replace several argument type from int to the correct one

For port-bind/unbind/bind-info commands, their arguments, such
as physical port (0 to 47), partition (0 to number of partitions -1),
logical port (1 to number of down stream port) number should be
nonnegative, nonnegative and positive. Change current CFG_INT type
to corresponding type accordingly.

Signed-off-by: wesley sheng <wesley.sheng@microchip.com>

wesleywesley committed:
Add description for timeout value argument

Add description of "(-1 forever)" to timeout value argument to
make clear about the acceptable value of timeout

Signed-off-by: wesley sheng <wesley.sheng@microchip.com>

wesleywesley committed:
Merge pull request #95 from Microsemi/cfg_int_fix

Replace several argument type from int to the correct one

wesleywesley committed:
Merge pull request #96 from Microsemi/timeout_desc

Add description for timeout value argument

kelvin-cao committed:
Merge pull request #92 from kelvin-cao/mrpc_errno

Enhance switchtec error handling to make it MRPC aware

kelvin-cao commented:
The final subcommand id has been determined, but this change in firmware is scheduled to about 2 months  later. Before that, we will need to release the software that match current firmware.  So I'd like to merge this into the microsemi repo, and later update the subcommand when firmware's change in place. Thought?

lsgunth commented:
Sounds fine to me.

kelvin-cao committed:
Merge pull request #98 from kelvin-cao/ums_reviewed_patches

Ums reviewed patches
